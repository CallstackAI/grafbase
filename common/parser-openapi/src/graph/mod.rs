use std::{borrow::Cow, collections::HashMap};

use dynaql::registry::resolvers::http::QueryParameterEncodingStyle;
use inflector::Inflector;
use openapiv3::StatusCode;
use petgraph::{
    graph::NodeIndex,
    visit::{EdgeRef, Reversed},
    Graph,
};

use crate::parsing::operations::OperationDetails;

mod enums;
mod input_object;
mod input_value;
mod operations;
mod output_type;
mod parameters;

pub use self::{
    enums::Enum,
    input_object::{InputField, InputObject},
    input_value::{InputValue, InputValueKind},
    operations::Operation,
    output_type::OutputType,
    parameters::{PathParameter, QueryParameter, RequestBody},
};

/// A graph representation of an OpenApi schema.
///
/// Can be queried to determine what resources are linked to what models etc.
pub struct OpenApiGraph {
    graph: Graph<Node, Edge>,
    operation_indices: Vec<NodeIndex>,
    pub metadata: crate::ApiMetadata,
}

impl OpenApiGraph {
    pub fn new(parsed: crate::parsing::Context, metadata: crate::ApiMetadata) -> Self {
        OpenApiGraph {
            graph: parsed.graph,
            operation_indices: parsed.operation_indices,
            metadata,
        }
    }
}

pub struct SchemaDetails {
    openapi_name: String,
    openapi: openapiv3::Schema,
}

impl SchemaDetails {
    pub fn new(openapi_name: String, openapi: openapiv3::Schema) -> Self {
        SchemaDetails { openapi_name, openapi }
    }
}

pub enum Node {
    /// A schema in the OpenApi spec.
    Schema(SchemaDetails),

    /// An individual HTTP operation in the OpenApi spec.
    Operation(OperationDetails),

    /// A GraphQL Object that may be needed in the output.
    Object,
    /// A scalar
    Scalar(ScalarKind),

    /// A union type that may be needed in the output.
    Union,

    /// An enum type
    Enum { values: Vec<String> },
}

#[derive(Debug)]
#[allow(clippy::enum_variant_names)]
pub enum Edge {
    /// Links an object with the types of it's fields.
    HasField { name: String, wrapping: WrappingType },

    /// The edge between a schema and its underlying type
    HasType { wrapping: WrappingType },

    /// An edge between an operation and the type/schema of one of its path parameters
    HasPathParameter { name: String, wrapping: WrappingType },

    /// An edge between an operation and the type/schema of one of its query parameters
    HasQueryParameter {
        name: String,
        wrapping: WrappingType,
        encoding_style: QueryParameterEncodingStyle,
    },

    /// An edge bewteen an operation and it's request type
    HasRequestType {
        content_type: RequestBodyContentType,
        wrapping: WrappingType,
    },

    /// An edge bewteen an operation and it's response type
    HasResponseType {
        status_code: StatusCode,
        #[allow(dead_code)]
        content_type: String,
        wrapping: WrappingType,
    },

    /// An edge between a union and it's constituent members
    HasUnionMember,
}

impl Node {
    fn as_operation(&self) -> Option<&OperationDetails> {
        match self {
            Node::Operation(op) => Some(op),
            _ => None,
        }
    }
}

#[derive(Debug)]
pub enum ScalarKind {
    String,
    Integer,
    Float,
    Boolean,
    #[allow(dead_code)]
    Id,
    JsonObject,
}

impl ScalarKind {
    fn type_name(&self) -> String {
        use dynaql::registry::scalars::{JSONScalar, SDLDefinitionScalar};

        match self {
            ScalarKind::String => "String".to_string(),
            ScalarKind::Integer => "Int".to_string(),
            ScalarKind::Float => "Float".to_string(),
            ScalarKind::Boolean => "Boolean".to_string(),
            ScalarKind::Id => "ID".to_string(),
            ScalarKind::JsonObject => JSONScalar::name().expect("JSONScalar to have a name").to_owned(),
        }
    }
}

impl std::fmt::Debug for Node {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Schema(schema) => f
                .debug_struct("Schema")
                .field("name", &schema.openapi_name)
                .finish_non_exhaustive(),
            Self::Operation(details) => f.debug_tuple("Operation").field(details).finish(),
            Self::Object => write!(f, "Object"),
            Self::Scalar(kind) => f.debug_tuple("Scalar").field(kind).finish(),
            Self::Enum { values } => f.debug_struct("Enum").field("values", values).finish(),
            Self::Union => write!(f, "Union"),
        }
    }
}

// The GraphQL spec calls the "NonNull"/"List" types "wrapping types" so I'm borrowing
// that terminology here
#[derive(Clone, Debug, PartialEq)]
pub enum WrappingType {
    NonNull(Box<WrappingType>),
    List(Box<WrappingType>),
    Named,
}

impl WrappingType {
    pub(super) fn wrap_list(self) -> WrappingType {
        WrappingType::List(Box::new(self))
    }

    pub(super) fn wrap_required(self) -> WrappingType {
        if matches!(self, WrappingType::NonNull(_)) {
            // Don't double wrap things in required
            self
        } else {
            WrappingType::NonNull(Box::new(self))
        }
    }

    pub(super) fn wrap_with(self, other: WrappingType) -> WrappingType {
        match other {
            WrappingType::NonNull(other_inner) => {
                if matches!(self, WrappingType::NonNull(_)) {
                    // Don't double wrap in required
                    self.wrap_with(*other_inner)
                } else {
                    WrappingType::NonNull(Box::new(self.wrap_with(*other_inner)))
                }
            }
            WrappingType::List(other_inner) => WrappingType::List(Box::new(self.wrap_with(*other_inner))),
            WrappingType::Named => self,
        }
    }

    pub fn contains_list(&self) -> bool {
        match self {
            WrappingType::NonNull(inner) => inner.contains_list(),
            WrappingType::List(_) => true,
            WrappingType::Named => false,
        }
    }
}

impl OpenApiGraph {
    fn type_name(&self, node: NodeIndex) -> Option<String> {
        match &self.graph[node] {
            schema @ Node::Schema { .. } => Some(schema.name()?),
            Node::Operation(_) => None,
            Node::Object | Node::Enum { .. } => {
                // OpenAPI objects are generally anonymous so we walk back up the graph to the
                // nearest named thing, and construct a name based on the fields in-betweeen.
                // Not ideal, but the best we can do.
                let reversed_graph = Reversed(&self.graph);

                let (_, mut path) = petgraph::algo::astar(
                    &reversed_graph,
                    node,
                    |current_node| self.graph[current_node].name().is_some(),
                    |_| 0,
                    |_| 0,
                )?;

                let named_node = path.pop()?;

                // Reverse our path so we can look things up in the original graph.
                path.reverse();

                let mut name_components = Vec::new();
                let mut path_iter = path.into_iter().peekable();
                while let Some(src_node) = path_iter.next() {
                    let Some(&dest_node) = path_iter.peek() else { break; };

                    // I am sort of assuming there's only one edge here.
                    // Should be the case at the moment but might need to update this to a loop if that changes
                    let edge = self.graph.edges_connecting(src_node, dest_node).next().unwrap();
                    if let Edge::HasField { name, .. } = edge.weight() {
                        name_components.push(name.as_str());
                    }
                }

                let root_name = self.graph[named_node].name().unwrap();
                name_components.push(root_name.as_str());

                name_components.reverse();
                Some(name_components.join("_").to_pascal_case())
            }
            Node::Scalar(kind) => Some(kind.type_name()),
            Node::Union => {
                // Unions are named based on the names of their constituent types.
                // Although it's perfectly possible for any of the members to be un-named
                // so this will probably require a bit more work at some point.
                let mut name = self
                    .graph
                    .edges(node)
                    .filter_map(|edge| match edge.weight() {
                        Edge::HasUnionMember => OutputType::from_index(edge.target(), self)?.name(self),
                        _ => None,
                    })
                    .collect::<Vec<_>>()
                    .join("Or");
                name.push_str("Union");
                Some(name)
            }
        }
    }
}

impl Node {
    // Used to determine whether this specific node type has a name.
    // To generate the full name of a particular node you should use the OpenApiGraph::type_name
    // function.
    fn name(&self) -> Option<String> {
        match self {
            Node::Schema(schema) => Some(
                // There's a title property that we _could_ use for a name, but the spec doesn't
                // enforce that it's unique and (certainly in stripes case) it is not.
                // Might do some stuff to work around htat, but for now it's either "x-resourceId"
                // which stripe use or the name of the schema in components.
                schema
                    .openapi
                    .schema_data
                    .extensions
                    .get("x-resourceId")
                    .and_then(|v| v.as_str())
                    .unwrap_or(schema.openapi_name.as_str())
                    .to_pascal_case(),
            ),
            Node::Operation(op) => op.operation_id.clone(),
            _ => None,
        }
    }
}

pub struct FieldName<'a>(Cow<'a, str>);

impl<'a> std::fmt::Display for FieldName<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let name = self.0.to_camel_case();

        write!(f, "{name}")
    }
}

#[derive(Clone, Debug)]
pub enum RequestBodyContentType {
    Json,
    FormEncoded(HashMap<String, QueryParameterEncodingStyle>),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    #[allow(clippy::redundant_clone)]
    fn test_wrap_with() {
        let named_type = WrappingType::Named;
        let list = named_type.clone().wrap_list();
        let required = named_type.clone().wrap_required();

        let required_list = list.clone().wrap_required();

        assert_eq!(&named_type.clone().wrap_with(list.clone()), &list);
        assert_eq!(&named_type.clone().wrap_with(required.clone()), &required);

        assert_eq!(&named_type.clone().wrap_with(required_list.clone()), &required_list);
        assert_eq!(
            &list.clone().wrap_with(required_list.clone()),
            &list.wrap_list().wrap_required()
        );

        // Check that we can't double wrap things in required
        assert_eq!(required.clone().wrap_with(required.clone()), required);
    }
}
