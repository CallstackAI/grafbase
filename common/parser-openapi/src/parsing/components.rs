use std::{collections::BTreeMap, rc::Rc};

use openapiv3::ReferenceOr;

use crate::Error;

use super::{operations::RequestBody, Context};

/// Re-usable components that can be referenced in an OpenAPI schema.
#[derive(Default)]
pub struct Components {
    pub responses: BTreeMap<Ref, Vec<ResponseComponent>>,
    pub request_bodies: BTreeMap<Ref, Rc<Vec<RequestBody>>>,
}

pub struct ResponseComponent {
    pub schema: Option<ReferenceOr<openapiv3::Schema>>,
    pub content_type: String,
}

impl Components {
    pub fn extend(&mut self, ctx: &mut Context, components: &openapiv3::Components) {
        for (name, response) in &components.responses {
            let Some(response) = response.as_item() else {
                ctx.errors.push(Error::TopLevelResponseWasReference(name.clone()));
                continue;
            };
            self.responses.insert(Ref::response(name), convert_response(response));
        }

        for (name, request_body) in &components.request_bodies {
            let Some(request_body) = request_body.as_item() else {
                ctx.errors.push(Error::TopLevelRequestBodyWasReference(name.clone()));
                continue;
            };
            self.request_bodies
                .insert(Ref::response(name), Rc::new(RequestBody::from_openapi(request_body)));
        }
    }
}

fn convert_response(response: &openapiv3::Response) -> Vec<ResponseComponent> {
    response
        .content
        .iter()
        .map(|(content_type, content)| ResponseComponent {
            schema: content.schema.clone(),
            content_type: content_type.clone(),
        })
        .collect()
}

#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Ref(String);

impl std::fmt::Display for Ref {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl Ref {
    pub fn absolute(absolute: &str) -> Ref {
        Ref(absolute.to_string())
    }

    pub fn schema(name: &str) -> Ref {
        Ref(format!("#/components/schemas/{name}"))
    }

    pub fn response(name: &str) -> Ref {
        Ref(format!("#/components/responses/{name}"))
    }

    pub fn request_body(name: &str) -> Ref {
        Ref(format!("#/components/request_bodies/{name}"))
    }
}
