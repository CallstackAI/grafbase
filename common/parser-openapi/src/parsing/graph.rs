use dynaql::registry::resolvers::http::{QueryParameterEncodingStyle, RequestBodyContentType};
use inflector::Inflector;
use once_cell::sync::Lazy;
use openapiv3::{AdditionalProperties, ReferenceOr, StatusCode, Type};
use petgraph::graph::NodeIndex;
use regex::Regex;

use crate::{
    graph::{ScalarKind, SchemaDetails, WrappingType},
    parsing::{
        components::{Components, Ref},
        operations::OperationDetails,
    },
    Error,
};

use super::{Context, Edge, Node};

pub fn extract_components(ctx: &mut Context, components: &openapiv3::Components) {
    for (name, schema) in &components.schemas {
        // I'm just going to assume that a top-level schema won't be a reference for now.
        // I think the only case where a user would do that is to reference a remote schema,
        // which is a PITA to support so lets not do that right now.
        let Some(schema) = schema.as_item() else {
            ctx.errors.push(Error::TopLevelSchemaWasReference(name.clone()));
            continue;
        };

        let index = ctx
            .graph
            .add_node(Node::Schema(SchemaDetails::new(name.clone(), schema.clone())));

        ctx.schema_index.insert(Ref::schema(name), index);
    }

    // Now we want to extract the spec for each of these schemas into our graph
    for (name, schema) in &components.schemas {
        extract_types(ctx, schema, ParentNode::Schema(ctx.schema_index[&Ref::schema(name)]));
    }
}

pub fn extract_operations(ctx: &mut Context, paths: &openapiv3::Paths, components: Components) {
    for (path, item) in &paths.paths {
        // Also going to assume that paths can't be references for now
        let Some(item) = item.as_item() else {
            ctx.errors.push(Error::TopLevelPathWasReference(path.clone()));
            continue;
        };

        for (method, operation) in item.iter() {
            let Ok(method) = method.parse() else {
                ctx.errors.push(Error::UnknownHttpMethod(method.to_string()));
                continue;
            };

            let operation = match OperationDetails::new(path.clone(), method, operation, &components) {
                Ok(operation) => operation,
                Err(e) => {
                    ctx.errors.push(e);
                    continue;
                }
            };
            let operation_index = ctx.graph.add_node(Node::Operation(operation.clone()));

            for parameter in operation.path_parameters {
                let parent = ParentNode::PathParameter {
                    name: parameter.name,
                    operation_index,
                };
                match parameter.schema {
                    Some(schema) => extract_types(ctx, &schema, parent),
                    None => {
                        // If the parameter has no schema we just assume it's a string.
                        ctx.add_type_node(parent, Node::Scalar(ScalarKind::String), false);
                    }
                }
            }

            for parameter in operation.query_parameters {
                let parent = ParentNode::QueryParameter {
                    name: parameter.name,
                    operation_index,
                    encoding_style: parameter.encoding_style,
                    required: parameter.required,
                };
                match parameter.schema {
                    Some(schema) => extract_types(ctx, &schema, parent),
                    None => {
                        // If the parameter has no schema we just assume it's a string.
                        ctx.add_type_node(parent, Node::Scalar(ScalarKind::String), false);
                    }
                }
            }

            for response in operation.responses {
                let Some(schema) = &response.schema else {
                    ctx.errors.push(Error::OperationMissingResponseSchema(operation.operation_id.clone().unwrap_or_else(|| format!("HTTP {method:?} {path}"))));
                    continue;
                };

                extract_types(
                    ctx,
                    schema,
                    ParentNode::OperationResponse {
                        status_code: response.status_code,
                        content_type: response.content_type,
                        operation_index,
                    },
                );
            }

            for request in operation.request_bodies.iter() {
                let Some(schema) = &request.schema else {
                    ctx.errors.push(Error::OperationMissingRequestSchema(operation.operation_id.clone().unwrap_or_else(|| format!("HTTP {method:?} {path}"))));
                    continue;
                };
                extract_types(
                    ctx,
                    schema,
                    ParentNode::OperationRequest {
                        content_type: request.content_type.clone(),
                        operation_index,
                        required: request.required,
                    },
                );
            }

            ctx.operation_indices.push(operation_index);
        }
    }
}

enum ParentNode {
    Schema(NodeIndex),
    OperationRequest {
        content_type: RequestBodyContentType,
        operation_index: NodeIndex,
        required: bool,
    },
    OperationResponse {
        status_code: StatusCode,
        content_type: String,
        operation_index: NodeIndex,
    },
    Field {
        object_index: NodeIndex,
        field_name: String,
        // Whether the field is required (which is a separate concept from nullable)
        required: bool,
    },
    List {
        nullable: bool,
        parent: Box<ParentNode>,
    },
    Union(NodeIndex),
    PathParameter {
        name: String,
        operation_index: NodeIndex,
    },
    QueryParameter {
        name: String,
        operation_index: NodeIndex,
        encoding_style: QueryParameterEncodingStyle,
        required: bool,
    },
}

impl Context {
    fn add_type_node(&mut self, parent: ParentNode, node: Node, nullable: bool) -> NodeIndex {
        let dest_index = self.graph.add_node(node);
        self.add_type_edge(parent, dest_index, nullable);
        dest_index
    }

    fn add_type_edge(&mut self, parent: ParentNode, dest_index: NodeIndex, nullable: bool) {
        let src_index = parent.node_index();
        let mut wrapping = WrappingType::Named;
        if !nullable {
            wrapping = wrapping.wrap_required();
        }
        self.graph
            .add_edge(src_index, dest_index, parent.create_edge_weight(wrapping));
    }
}

impl ParentNode {
    fn node_index(&self) -> NodeIndex {
        match self {
            ParentNode::Union(idx) | ParentNode::Schema(idx) => *idx,
            ParentNode::OperationResponse { operation_index, .. }
            | ParentNode::OperationRequest { operation_index, .. }
            | ParentNode::PathParameter { operation_index, .. }
            | ParentNode::QueryParameter { operation_index, .. } => *operation_index,
            ParentNode::Field { object_index, .. } => *object_index,
            ParentNode::List { parent, .. } => parent.node_index(),
        }
    }

    fn create_edge_weight(&self, wrapping: WrappingType) -> Edge {
        match self {
            ParentNode::Schema(_) => Edge::HasType { wrapping },
            ParentNode::OperationRequest {
                content_type, required, ..
            } => Edge::HasRequestType {
                content_type: content_type.clone(),
                // If a parameter is marked as not required, we need to make sure that we
                // don't record it as required, regardless of what the schema says.
                wrapping: wrapping.set_required(*required),
            },
            ParentNode::OperationResponse {
                status_code,
                content_type,
                ..
            } => Edge::HasResponseType {
                content_type: content_type.clone(),
                status_code: status_code.clone(),
                wrapping,
            },
            ParentNode::Field {
                field_name, required, ..
            } => Edge::HasField {
                name: field_name.clone(),
                // wrapping will have had the nullability of a field applied at this
                // point.  But OpenAPI schemas often don't bother specifying the
                // nullability of object fields and just use required, so we're better
                // off ignoring `nullable` and just relying on `required` here.
                wrapping: wrapping.set_required(*required),
            },
            ParentNode::List { nullable, parent } => {
                // Ok, so call parent.to_edge_weight and then modifiy the wrapping in it.
                // Wrapping the wrapping in a List(Required()) or just List() as appropriate.
                let mut wrapping = wrapping.wrap_list();
                if !nullable {
                    wrapping = wrapping.wrap_required();
                }
                parent.create_edge_weight(wrapping)
            }
            ParentNode::Union { .. } => Edge::HasUnionMember,
            ParentNode::PathParameter { name, .. } => Edge::HasPathParameter {
                name: name.clone(),
                // Path parameters are always required, so lets make sure they are here too.
                wrapping: wrapping.wrap_required(),
            },
            ParentNode::QueryParameter {
                name,
                encoding_style,
                required,
                ..
            } => Edge::HasQueryParameter {
                name: name.clone(),
                // If a parameter is marked as not required, we need to make sure that we
                // don't record it as required, regardless of what the schema says.
                wrapping: wrapping.set_required(*required),
                encoding_style: *encoding_style,
            },
        }
    }
}

fn extract_types(ctx: &mut Context, schema_or_ref: &ReferenceOr<openapiv3::Schema>, parent: ParentNode) {
    use openapiv3::SchemaKind;

    match schema_or_ref {
        ReferenceOr::Reference { reference } => {
            let reference = Ref::absolute(reference);
            let Some(schema) = ctx.schema_index.get(&reference) else {
                ctx.errors.push(Error::UnresolvedReference(reference));
                return;
            };

            ctx.add_type_edge(parent, *schema, false);
        }
        ReferenceOr::Item(schema) => match &schema.schema_kind {
            SchemaKind::Type(Type::String(ty)) => {
                if ty.enumeration.is_empty() || !ty.enumeration.iter().all(is_valid_enum_value) {
                    ctx.add_type_node(parent, Node::Scalar(ScalarKind::String), schema.schema_data.nullable);
                } else {
                    ctx.add_type_node(
                        parent,
                        Node::Enum {
                            values: ty.enumeration.iter().flatten().cloned().collect(),
                        },
                        schema.schema_data.nullable,
                    );
                }
            }
            SchemaKind::Type(Type::Boolean {}) => {
                ctx.add_type_node(parent, Node::Scalar(ScalarKind::Boolean), schema.schema_data.nullable);
            }
            SchemaKind::Type(Type::Integer(_)) => {
                ctx.add_type_node(parent, Node::Scalar(ScalarKind::Integer), schema.schema_data.nullable);
            }
            SchemaKind::Type(Type::Number(_)) => {
                ctx.add_type_node(parent, Node::Scalar(ScalarKind::Float), schema.schema_data.nullable);
            }
            SchemaKind::Type(Type::Object(obj)) => {
                if obj.properties.is_empty() {
                    // If the object is empty _and_ there's no additionalProperties we don't bother
                    // emiting an object for it.  Not sure if this is a good idea - could be some APIs
                    // that _require_ an empty object.  But lets see what happens
                    if obj.additional_properties != Some(AdditionalProperties::Any(false)) {
                        ctx.add_type_node(parent, Node::Scalar(ScalarKind::JsonObject), false);
                    }
                    return;
                }
                let object_index = ctx.add_type_node(parent, Node::Object, schema.schema_data.nullable);
                for (field_name, field_schema_or_ref) in &obj.properties {
                    let required = obj.required.contains(field_name);
                    extract_types(
                        ctx,
                        &field_schema_or_ref.clone().unbox(),
                        ParentNode::Field {
                            object_index,
                            field_name: field_name.clone(),
                            required,
                        },
                    );
                }
            }
            SchemaKind::Type(Type::Array(arr)) => {
                let Some(items) = arr.items.clone() else {
                    // We don't support array without items, so error if it's missing
                    ctx.errors.push(Error::ArrayWithoutItems);
                    return;
                };

                extract_types(
                    ctx,
                    &items.unbox(),
                    ParentNode::List {
                        nullable: schema.schema_data.nullable,
                        parent: Box::new(parent),
                    },
                );
            }
            SchemaKind::OneOf { one_of: schemas } | SchemaKind::AnyOf { any_of: schemas } => {
                if schemas.len() == 1 {
                    // The Stripe API has anyOfs containing a single item.
                    // For ease of use we simplify this to just point direct at the underlying type.
                    extract_types(ctx, schemas.first().unwrap(), parent);
                    return;
                }

                let union_index = ctx.add_type_node(parent, Node::Union, schema.schema_data.nullable);
                for schema in schemas {
                    extract_types(ctx, schema, ParentNode::Union(union_index));
                }
            }
            SchemaKind::AllOf { .. } => {
                ctx.errors.push(Error::AllOfSchema);
            }
            SchemaKind::Not { .. } => {
                ctx.errors.push(Error::NotSchema);
            }
            SchemaKind::Any(any) => {
                // We treat an any very similar to an object
                if any.properties.is_empty() {
                    // If there's no explicit properties we make this a custom scalar
                    ctx.add_type_node(parent, Node::Scalar(ScalarKind::JsonObject), false);
                    return;
                }
                let object_index = ctx.add_type_node(parent, Node::Object, schema.schema_data.nullable);
                for (field_name, field_schema_or_ref) in &any.properties {
                    let required = any.required.contains(field_name);
                    extract_types(
                        ctx,
                        &field_schema_or_ref.clone().unbox(),
                        ParentNode::Field {
                            object_index,
                            field_name: field_name.clone(),
                            required,
                        },
                    );
                }
            }
        },
    }
}

// OpenAPI enums can be basically any string, but we're much more limited
/// in GraphQL.  This checks if this value is valid in GraphQL or not.
fn is_valid_enum_value(value: &Option<String>) -> bool {
    static REGEX: Lazy<Regex> = Lazy::new(|| Regex::new("^[A-Z_][A-Z0-9_]*$").unwrap());
    value
        .as_deref()
        .map(|value| value.to_screaming_snake_case())
        .filter(|value| REGEX.is_match(value))
        .is_some()
}
