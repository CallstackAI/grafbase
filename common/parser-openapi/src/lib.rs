use dynaql::registry::Registry;
use graph::OpenApiGraph;
use openapiv3::OpenAPI;
use parsing::components::Ref;
use url::Url;

mod graph;
mod output;
mod parsing;
mod validation;

pub fn parse_spec(
    data: &str,
    format: Format,
    mut metadata: ApiMetadata,
    registry: &mut Registry,
) -> Result<(), Vec<Error>> {
    // Make sure we have a trailing slash on metadata so that Url::join works correctly.
    ensure_trailing_slash(&mut metadata.url).map_err(|_| vec![Error::InvalidUrl(metadata.url.to_string())])?;

    let spec = match format {
        Format::Json => serde_json::from_str::<OpenAPI>(data).map_err(|e| vec![Error::JsonParsingError(e)])?,
        Format::Yaml => serde_yaml::from_str::<OpenAPI>(data).map_err(|e| vec![Error::YamlParsingError(e)])?,
    };

    let graph = OpenApiGraph::new(parsing::parse(spec)?, metadata);

    validation::validate(&graph)?;

    output::output(&graph, registry);

    Ok(())
}

pub struct ApiMetadata {
    pub name: String,
    pub url: Url,
}

pub enum Format {
    Json,
    Yaml,
}

impl Format {
    pub fn guess(content_type: Option<&str>, url: &str) -> Self {
        if let Some(content_type) = content_type {
            if content_type == "application/json" {
                return Format::Json;
            }
            if content_type.contains("yaml") {
                return Format::Yaml;
            }
        }
        if let Some(extension) = extract_extension(url) {
            if extension.eq_ignore_ascii_case("json") {
                return Format::Json;
            }
            if extension.eq_ignore_ascii_case("yml") || extension.eq_ignore_ascii_case("yaml") {
                return Format::Yaml;
            }
        }

        // YAML is a superset of JSON so lets just fallback to parsing as YAML.
        Format::Yaml
    }
}

fn extract_extension(url: &str) -> Option<String> {
    let url = Url::parse(url).ok()?;
    let last_segment = url.path_segments()?.last()?;

    let extension = std::path::Path::new(last_segment).extension()?;

    Some(extension.to_str()?.to_string())
}

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("Could not parse the open api spec: {0}")]
    JsonParsingError(serde_json::Error),
    #[error("Could not parse the open api spec: {0}")]
    YamlParsingError(serde_yaml::Error),
    #[error("The schema component {0} was a reference, which we don't currently support.")]
    TopLevelSchemaWasReference(String),
    #[error("The path component {0} was a reference, which we don't currently support.")]
    TopLevelPathWasReference(String),
    #[error("The response component {0} was a reference, which we don't currently support.")]
    TopLevelResponseWasReference(String),
    #[error("The request body component {0} was a reference, which we don't currently support.")]
    TopLevelRequestBodyWasReference(String),
    #[error("The parameter component {0} was a reference, which we don't currently support.")]
    TopLevelParameterWasReference(String),
    #[error("Couldn't parse HTTP verb: {0}")]
    UnknownHttpVerb(String),
    #[error("The operation {0} didn't have a response schema")]
    OperationMissingResponseSchema(String),
    #[error("The operation {0} didn't have a response schema")]
    OperationMissingRequestSchema(String),
    #[error("Encountered an array without items, which we don't currently support")]
    ArrayWithoutItems,
    #[error("Encountered a not schema, which we don't currently support")]
    NotSchema,
    #[error("Encountered an allOf schema, which we don't currently support")]
    AllOfSchema,
    #[error("Encountered an any schema, which we don't currently support")]
    AnySchema,
    #[error("Found a reference {0} which didn't seem to exist in the spec")]
    UnresolvedReference(Ref),
    #[error("Received an invalid URL: {0} ")]
    InvalidUrl(String),
    #[error("The path parameter {0} on operation {1} is an object, which is currently unsupported")]
    PathParameterIsObject(String, String),
    #[error("The path parameter {0} on operation {1} is a list, which is currently unsupported")]
    PathParameterIsList(String, String),
    #[error("The query parameter {0} on operation {1} is an object, which is currently unsupported")]
    QueryParameterIsObject(String, String),
    #[error("The query parameter {0} on operation {1} is a list, which is currently unsupported")]
    QueryParameterIsList(String, String),
    #[error("The query parameter {0} on operation {1} has a style {2}, which is currently unsupported")]
    UnsupportedQueryParameterStyle(String, String, String),
    #[error("The query parameter {0} on operation {1} has an object nested inside a list, which is unsupported")]
    ObjectNestedInsideListQueryParamter(String, String),
    #[error("The query parameter {0} on operation {1} has a non-scalar nested inside an object, which is unsupported")]
    NonScalarNestedInsideObjectQueryParameter(String, String),
    #[error("The query parameter {0} on operation {1} has a non-scalar nested inside an object, which is unsupported")]
    ListNestedInsideObjectQueryParameter(String, String),
}

fn is_ok(status: &openapiv3::StatusCode) -> bool {
    match status {
        openapiv3::StatusCode::Code(200) => true,
        openapiv3::StatusCode::Range(_range) => todo!(),
        _ => false,
    }
}

fn ensure_trailing_slash(url: &mut Url) -> Result<(), ()> {
    let mut segments = url.path_segments_mut()?;

    segments.pop_if_empty();
    segments.push("");

    Ok(())
}

#[cfg(test)]
mod tests {
    use dynaql::{indexmap::IndexMap, CacheControl};

    use super::*;

    #[test]
    fn test_stripe_output() {
        let spec = std::fs::read_to_string("test_data/stripe.openapi.json").unwrap();

        let mut registry = default_registry();

        parse_spec(&spec, Format::Json, metadata(), &mut registry).unwrap();

        insta::assert_snapshot!(registry.export_sdl(false));
    }

    #[test]
    fn test_petstore_output() {
        let spec = std::fs::read_to_string("test_data/petstore.openapi.json").unwrap();

        let mut registry = default_registry();

        parse_spec(&spec, Format::Json, metadata(), &mut registry).unwrap();

        insta::assert_snapshot!(registry.export_sdl(false));
        insta::assert_debug_snapshot!(registry);
    }

    fn metadata() -> ApiMetadata {
        ApiMetadata {
            name: "example".into(),
            url: Url::parse("http://example.com").unwrap(),
        }
    }

    fn default_registry() -> Registry {
        let mut registry = Registry {
            query_type: "Query".to_string(),
            ..Registry::default()
        };
        registry.types.insert(
            "Query".to_string(),
            dynaql::registry::MetaType::Object {
                name: "Query".to_string(),
                description: None,
                fields: IndexMap::new(),
                cache_control: CacheControl {
                    public: true,
                    max_age: 0,
                },
                extends: false,
                keys: None,
                visible: None,
                is_subscription: false,
                is_node: false,
                rust_typename: "Query".to_string(),
                constraints: vec![],
            },
        );

        registry
    }
}
