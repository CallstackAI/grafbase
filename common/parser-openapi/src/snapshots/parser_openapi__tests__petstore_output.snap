---
source: common/parser-openapi/src/lib.rs
expression: registry.export_sdl(false)
---
"""
A JSON Value
"""
scalar JSON
type Mutation {
	petstore: PetstoreMutations
}
type PetstoreApiResponse {
	message: String
	type: String
	code: Int
}
type PetstoreCategory {
	name: String
	id: Int
}
input PetstoreCategoryInput {
	name: String
	id: Int
}
enum PetstoreFindPetsByStatus {
	AVAILABLE
	PENDING
	SOLD
}
type PetstoreMutations {
	updatePet(input: PetstorePetInput!): PetstorePet!
	addPet(input: PetstorePetInput!): PetstorePet!
	uploadFile(petId: Int!, additionalMetadata: String): PetstoreApiResponse!
	placeOrder(input: PetstoreOrderInput!): PetstoreOrder!
	createUsersWithListInput(input: [PetstoreUserInput!]!): PetstoreUser!
}
type PetstoreOrder {
	complete: Boolean
	status: PetstoreOrderStatus
	shipDate: String
	quantity: Int
	petId: Int
	id: Int
}
input PetstoreOrderInput {
	complete: Boolean
	status: PetstoreOrderStatus
	shipDate: String
	quantity: Int
	petId: Int
	id: Int
}
enum PetstoreOrderStatus {
	PLACED
	APPROVED
	DELIVERED
}
type PetstorePet {
	status: PetstorePetStatus
	tags: [PetstoreTag!]
	photoUrls: [String!]!
	category: PetstoreCategory
	name: String!
	id: Int
}
input PetstorePetInput {
	status: PetstorePetStatus
	tags: [PetstoreTagInput!]!
	photoUrls: [String!]!
	category: PetstoreCategoryInput!
	name: String!
	id: Int
}
enum PetstorePetStatus {
	AVAILABLE
	PENDING
	SOLD
}
type PetstoreQueries {
	findPetsByStatus(status: PetstoreFindPetsByStatus): [PetstorePet!]!
	findPetsByTags(tags: [String!]): [PetstorePet!]!
	getPetById(petId: Int!): PetstorePet!
	getInventory: JSON!
	getOrderById(orderId: Int!): PetstoreOrder!
	loginUser(password: String, username: String): String!
	getUserByName(username: String!): PetstoreUser!
}
type PetstoreTag {
	name: String
	id: Int
}
input PetstoreTagInput {
	name: String
	id: Int
}
type PetstoreUser {
	userStatus: Int
	phone: String
	password: String
	email: String
	lastName: String
	firstName: String
	username: String
	id: Int
}
input PetstoreUserInput {
	userStatus: Int
	phone: String
	password: String
	email: String
	lastName: String
	firstName: String
	username: String
	id: Int
}
type Query {
	petstore: PetstoreQueries
}
schema {
	query: Query
	mutation: Mutation
}

